tools/database/src/analyse_gas_usage.rs

```diff
@@ -0,0 +1,348 @@
+        }
+        result
+    }
+
+    pub fn merge(&mut self, other: GasUsageStats) {
+        for (shard_uid, shard_usage) in other.shards {
+            self.add_gas_usage_in_shard(shard_uid, shard_usage);
+        }
+    }
+}
+
+fn get_gas_usage_in_block(
+    block: &Block,
+    chain_store: &ChainStore,
+    epoch_manager: &EpochManager,
+) -> GasUsageStats {
+    let block_info: Arc<BlockInfo> = epoch_manager.get_block_info(block.hash()).unwrap();
+    let epoch_id: &EpochId = block_info.epoch_id();
+    let shard_layout: ShardLayout = epoch_manager.get_shard_layout(epoch_id).unwrap();
+
+    let mut result = GasUsageStats::new();
+
+    // Go over every chunk in this block and gather data
+    for chunk_header in block.chunks().iter() {
+        let shard_id: ShardId = chunk_header.shard_id();
+        let shard_uid: ShardUId = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
+
+        let mut gas_usage_in_shard = GasUsageInShard::new();
+
+        // The outcome of each transaction and receipt executed in this chunk is saved in the database as an ExecutionOutcome.
+        // Go through all ExecutionOutcomes from this chunk and record the gas usage.
+        let outcome_ids: Vec<CryptoHash> =
+            chain_store.get_outcomes_by_block_hash_and_shard_id(block.hash(), shard_id).unwrap();
+        for outcome_id in outcome_ids {
+            let outcome: ExecutionOutcome = chain_store
+                .get_outcome_by_id_and_block_hash(&outcome_id, block.hash())
+                .unwrap()
+                .unwrap()
+                .outcome;
+
+            // Sanity check - make sure that the executor of this outcome belongs to this shard
+            let account_shard: ShardId =
+                account_id_to_shard_id(&outcome.executor_id, &shard_layout);
+            assert_eq!(account_shard, shard_id);
+
+            gas_usage_in_shard.add_used_gas(outcome.executor_id, outcome.gas_burnt.into());
+        }
+
+        result.add_gas_usage_in_shard(shard_uid, gas_usage_in_shard);
+    }
+
+    result
+}
+
+/// A struct that can be used to find N biggest accounts by gas usage in an efficient manner.
+struct BiggestAccountsFinder {
+    accounts: BTreeSet<(BigGas, AccountId)>,
+    accounts_num: usize,
+}
+
+impl BiggestAccountsFinder {
+    pub fn new(accounts_num: usize) -> BiggestAccountsFinder {
+        BiggestAccountsFinder { accounts: BTreeSet::new(), accounts_num }
+    }
+
+    pub fn add_account_stats(&mut self, account: AccountId, used_gas: BigGas) {
+        self.accounts.insert((used_gas, account));
+
+        // If there are more accounts than desired, remove the one with the smallest gas usage
+        while self.accounts.len() > self.accounts_num {
+            self.accounts.pop_first();
+        }
+    }
+
+    pub fn get_biggest_accounts(&self) -> impl Iterator<Item = (AccountId, BigGas)> + '_ {
+        self.accounts.iter().rev().map(|(gas, account)| (account.clone(), *gas))
```

Comment:
It also reverses the range so that the iterator starts at the biggest account instead of the smallest.
Then it also swaps the tuple, I think it looks nicer as `(AccountId, BigGas)` ¯\_(ツ)_/¯